### ISSUE TITLE
KH-9xx: needle 0 (L100) always gets selected to B when going to the left

### ISSUE THREAD

**Issue by jonathanperret** [jpcornil-git/ayabAsync-firmware#1]
* Version tested: ayabAsync 1.1.3, ayab-desktop 1.0 beta 3
* Using a 200-needles pattern like this one:
<img src="https://github.com/user-attachments/assets/e691a722-e913-41ea-a536-8ee67121b11e" width="400">

* All other settings at default.
* "Belt phase" should be "regular". On hardware, if the bug does not appear, try again with the carriage shifted by 8 needles on the belt.
* Knit one row left to right: needle L100 is selected to D as expected.
* Knit one row right to left: needle L100 is selected to B despite pattern having only black squares in that column.

This can be reproduced both on a KH910 and in `ayab-simavr`.

Tracing through the code, what seems to happen is that when `selectPosition` gets to `-1`, `Machine::solenoidToSet` is still called and returns a nonsensical solenoid index of `255`.
Because the belt shift is "regular", `_machine->solenoidShift` is not called and so we end up calling `_solenoids->reset(255)`.
This ends up doing `_states |= (1 << 255)`. But `1 << 255` is [undefined behavior](https://en.cppreference.com/w/cpp/language/operator_arithmetic) (shift by more bits than the target type contains), and in this environment it appears to always return `1` so we end up resetting the first solenoid.

**Comment by jpcornil-git** [comment]
He, he, nice one ! ... unlike e.g. python, c/c++ doesn't floor the result of an integer division to compute the remainder :-)
https://github.com/jpcornil-git/ayabAsync-firmware/blob/v1.0/lib/components/machine/machine.cpp#L58

Need to think about a clean way to fix this now

**Comment by jpcornil-git** [comment]
I would do the following:
``` c++
uint8_t Machine::solenoidToSet(int16_t needleToSet) {
  // Note: Remainder of a negative number is negative in c/c++ (quotient of the related division is truncated towards zero)
  uint8_t solenoid;
  switch (_type) {
    case MachineType::Kh270:
      solenoid = (uint8_t) ((needleToSet + 4) % 12);
      return (solenoid < 12)  ? solenoid : solenoid + 12; // see Note
    default:
      solenoid = (uint8_t) (needleToSet % 16);
      return (solenoid < 16)  ? solenoid : solenoid + 16; // see Note
  }
}
```

Thoughts ?

**Comment by jonathanperret** [comment]
Well first off I'm not sure we need to be setting solenoids for out-of-bed "needles".

But if we want to do it, I think the modular arithmetic could be clearer here — I'm not even sure it's correct in the case of the KH270, because 12 doesn't divide 256 so it's not obvious that it's safe to wrap around. Even if it can be proven safe, I wouldn't be comfortable with it.

I would probably instead introduce a function to do positive modulo, like e.g. this (from StackOverflow):

```
unsigned modulo( int value, unsigned m) {
    int mod = value % (int)m;
    if (mod < 0) {
        mod += m;
    }
    return mod;
}
```

**Comment by jpcornil-git** [comment]
A priori (with a correct modulo arithmetic :-)), it doesn't hurt as this would be the solenoid to set for a wider bed anyway.

The modulo operation of a negative number, because of the truncated division in c++,  gives you a remainder between -15 and 0 for % 16 (-11 and 0 for % 12) => adding 16 (or 12) bring you back to the same as for positive numbers, see below ("truncating" for c++, "flooring" for python/excel/...)

![image](https://github.com/user-attachments/assets/b93b03f7-fca2-4548-a9f0-048199676d24)

To avoid the wraparound with int to uint conversion, I could just do the following:

``` c++
uint8_t Machine::solenoidToSet(int16_t needleToSet) {
  // Note: Remainder of a negative number is negative in c/c++ (quotient of the related division is truncated towards zero)
  int16_t solenoid;
  switch (_type) {
    case MachineType::Kh270:
      solenoid =  (needleToSet + 4) % 12;
      return (solenoid >= 0)  ? (uint8_t) solenoid : (uint8_t) (solenoid + 12); // see Note
    default:
      solenoid = needleToSet % 16;
      return (solenoid >= 0)  ? (uint8_t) solenoid : (uint8_t) (solenoid + 16); // see Note
  }
}
```
An alternative approach would be to make L258-L293 of knitter.cpp conditional to selectPosition being in the needle bed but then I need to keep  _carriage->workFinished (L290) out with its associated machineSide logic while out of the needle bed, i.e. split into two sections, one handling solenoid selection and the other the carriage/_currentLine.finished logic.

**Comment by jonathanperret** [comment]
> A priori (with a correct modulo arithmetic :-)), it doesn't hurt as this would be the solenoid to set for a wider bed anyway.
> 
> The modulo operation of a negative number, because of the truncated division in c++, gives you a remainder between -15 and 0 for % 16 (-11 and 0 for % 12) => adding 16 (or 12) bring you back to the same as for positive numbers, see below ("truncating" for c++, "flooring" for python/excel/...)

I'm familiar with how the modulo operation behaves in C++ vs Python, hence why I suggested a "positive modulo" implementation (like Python) as a separate to get the result we need while not obscuring the computation in `solenoidToSet` with arithmetic details. Also, my doubts about correctness were not around converting C++'s modulo operation to Python's but about using unsigned integers to do signed arithmetic, which is another kettle of fish I'd rather avoid entirely. I think this is clearer but it's up to you:

```c++
uint8_t Machine::solenoidToSet(int16_t needleToSet) {
  switch (_type) {
    case MachineType::Kh270:
      return positiveModulo(needleToSet + 4, 12);
    default:
      return positiveModulo(needleToSet, 16);
  }
}
```

> An alternative approach would be to make L258-L293 of knitter.cpp conditional to selectPosition being in the needle bed but then I need to keep _carriage->workFinished (L290) out with its associated machineSide logic while out of the needle bed, i.e. split into two sections, one handling solenoid selection and the other the carriage/_currentLine.finished logic.

Yes, that was the first thing I tried and of course nothing worked when I made everything conditional on the needle existing, so the code got a bit complicated.

So if it turns out simpler to always set a solenoid even though there's no needle being driven at that point, I think it's fine to do it that way.

**Comment by jpcornil-git** [comment]
I like that approach as well, very readable

``` c++

int16_t positiveModulo(int16_t dividend, int16_t divisor) {
  int16_t remainder = dividend % divisor;
  if (remainder < 0) {
    remainder += divisor;
  }
  return remainder;
}

uint8_t Machine::solenoidToSet(int16_t needleToSet) {
  switch (_type) {
    case MachineType::Kh270:
      return (uint8_t) positiveModulo(needleToSet + 4, 12);
    default:
      return (uint8_t) positiveModulo(needleToSet, 16);
  }
}
```

**Comment by jpcornil-git** [comment]
Fixed, see 60fa8a3
